pkg/fs/remote/gcp/gcs.go

```diff
@@ -0,0 +1,250 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Package gcp provides the GCS file system.
+package gcp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+
+	"github.com/apache/skywalking-banyandb/pkg/fs/remote"
+	"github.com/apache/skywalking-banyandb/pkg/fs/remote/checksum"
+	config2 "github.com/apache/skywalking-banyandb/pkg/fs/remote/config"
+)
+
+var _ remote.FS = (*gcsFS)(nil)
+
+// gcsFS implements remote.FS backed by Google Cloud Storage.
+// Field order is optimized to reduce struct padding.
+type gcsFS struct {
+	client   *storage.Client
+	verifier checksum.Verifier
+	bucket   string
+	basePath string
+}
+
+// NewFS creates a new GCS-backed FS. The input path must be in the form
+// <bucket>/<optional-basePath> (without leading slash).
+func NewFS(p string, cfg *config2.FsConfig) (remote.FS, error) {
+	if cfg == nil {
+		cfg = &config2.FsConfig{}
+	}
+	if cfg.GCP == nil {
+		cfg.GCP = &config2.GCPConfig{}
+	}
+
+	rawPath := strings.Trim(p, "/")
+
+	var bucket, basePath string
+
+	if cfg.GCP.Bucket != "" {
+		bucket = cfg.GCP.Bucket
+		if strings.HasPrefix(rawPath, bucket+"/") {
+			basePath = strings.TrimPrefix(rawPath, bucket+"/")
+		} else {
+			basePath = rawPath
+		}
+	} else {
+		parts := strings.SplitN(rawPath, "/", 2)
+		if len(parts) == 0 || parts[0] == "" {
+			return nil, fmt.Errorf("bucket name must be specified either in config or path")
+		}
+		bucket = parts[0]
+		cfg.GCP.Bucket = bucket
+		if len(parts) == 2 {
+			basePath = parts[1]
+		} else {
+			basePath = ""
+		}
+	}
+
+	ctx := context.Background()
+	var client *storage.Client
+	var err error
+	if cfg.GCP.GCPServiceAccountFile != "" {
+		client, err = storage.NewClient(ctx, option.WithCredentialsFile(cfg.GCP.GCPServiceAccountFile))
```

Comment:
No file existence check. No permission validation, the mode of this file should be 600 since it's a credential.